### PR TITLE
feat: Ecwid Updates

### PR DESCRIPTION
**Pull request by trendoscope-algorithms**
Updates related to ECWID plugin

- modified app to use mjs
- added new source get-paid-orders which runs based on scheduled poll

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4163"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thank you so much for submitting this! We've added it to our backlog to review, and our team has been notified.

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

File: components/ecwid/ecwid.app.mjs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        method: "GET",
+        headers: {
+          "user-agent": "@PipedreamHQ/pipedream v0.1",
+          "accept": "application/json",
+          "Authorization": `Bearer ${this.$auth.oauth_access_token}`,
+        },
+      });
+      console.log(res);
+      return res.data.items;
+    },
```
Comment (lcaresia) [inline components/ecwid/ecwid.app.mjs:27]: Move all this to a method like this: https://github.com/PipedreamHQ/pipedream/blob/c149db25eac78922f9c13e9b249f11576cda513b/components/awork/awork.app.mjs#L14

File: components/ecwid/ecwid.app.mjs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    authKeys() {
+      console.log(Object.keys(this.$auth));
+    },
+
+    async getOrders(history = 30, paymentStatus = "PAID", fulfilmentStatus = "AWAITING_PROCESSING") {
+      const fromDate = new Date();
+      fromDate.setDate(fromDate.getDate() - history);
+      const fromDateTimeStamp = Math.floor(fromDate.getTime() / 1000).toString();
+      const res = await axios({
+        url: `https://app.ecwid.com/api/v3/${this.$auth.storeId}/orders?` +
```
Comment (lcaresia) [inline components/ecwid/ecwid.app.mjs:16]: Move URL to a getter method.

```suggestion
        url: `${this._apiUrl()}${this.$auth.storeId}/orders?` +
```

File: components/ecwid/package.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "name": "@pipedream/ecwid",
   "version": "0.0.1",
   "description": "Pipedream ecwid Components",
```
Comment (lcaresia) [inline components/ecwid/package.json:4]:
```suggestion
  "description": "Pipedream Ecwid Components",
```

File: components/ecwid/sources/get-paid-orders/get-paid-orders.mjs [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      label: "Set fulfilment status to PROCESSING (Preferred)",
+      description: "Upon receiving the order, fulfilment status will be set to PROCESSING so that " +
+          "the same order is not fetched again. If unchecked, it needs to be handled in the workflow",
+      type: "boolean",
+      default: true,
+    },
+    ecwid,
+  },
+  type: "source",
+  methods: {},
```
Comment (lcaresia) [inline components/ecwid/sources/get-paid-orders/get-paid-orders.mjs:32]: Try to use deploy hook to best experience.

https://pipedream.com/docs/components/api/#deploy

```suggestion
  methods: {},
  hooks: {
    deploy() {
      // deploy code here
    },
  },
```

File: components/ecwid/ecwid.app.mjs [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        headers: {
+          "user-agent": "@PipedreamHQ/pipedream v0.1",
+          "accept": "application/json",
+          "Authorization": `Bearer ${this.$auth.oauth_access_token}`,
+        },
+        data: {
+          "fulfillmentStatus": fulfilmentStatus,
+        },
+      });
+    },
```
Comment (lcaresia) [inline components/ecwid/ecwid.app.mjs:41]: The same of above here.

File: components/ecwid/sources/get-paid-orders/get-paid-orders.mjs (hunk 4 above)
Comment (trendoscope-algorithms) [reply components/ecwid/sources/get-paid-orders/get-paid-orders.mjs:32]: At present, we don't need a webhook. Ecwid uses a centralised webhook for entire app and registration of webhook is manual (By requesting via customer service). It does not really fit with creating webhook per deployed components.

File: components/ecwid/ecwid.app.mjs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      fromDate.setDate(fromDate.getDate() - history);
+      const fromDateTimeStamp = Math.floor(fromDate.getTime() / 1000).toString();
+      const res = await axios({
+        url: `https://app.ecwid.com/api/v3/${this.$auth.storeId}/orders?` +
+              `paymentStatus=${paymentStatus}&fulfillmentStatus=${fulfilmentStatus}&createdFrom=${fromDateTimeStamp}`,
+        method: "GET",
+        headers: {
+          "user-agent": "@PipedreamHQ/pipedream v0.1",
+          "accept": "application/json",
+          "Authorization": `Bearer ${this.$auth.oauth_access_token}`,
```
[the remark this answers is not in the document]
Comment (trendoscope-algorithms) [reply components/ecwid/ecwid.app.mjs]: Done

File: components/ecwid/ecwid.app.mjs (hunk 1 above)
Comment (trendoscope-algorithms) [reply components/ecwid/ecwid.app.mjs]: Done.

File: components/ecwid/ecwid.app.mjs (hunk 5 above)
Comment (trendoscope-algorithms) [reply components/ecwid/ecwid.app.mjs]: Done.

File: components/ecwid/sources/get-paid-orders/get-paid-orders.mjs (hunk 4 above)
Comment (feyzullah) [reply components/ecwid/sources/get-paid-orders/get-paid-orders.mjs:40]: Hi @trendoscope-algorithms, @lcaresia, as I understand, historical records are already emitted here. So I think that we do not need `deploy` hook. However, if I get it correctly, whenever the timer ticks this one is going to emit all related orders again. We can avoid this by setting a db record and check it before making the request. Please check the db docs [here](https://pipedream.com/docs/components/api/#db) and an example [here](https://github.com/PipedreamHQ/pipedream/blob/f31508b6fa735a0b3ebafccbd84e5f86028ec6ef/components/pinterest/sources/common.mjs#L16). As I see [here](https://api-docs.ecwid.com/reference/search-orders#orderentry), there are `createDate` and `updateDate` props in `OrderEntry`, I think these can help while getting new or updated orders.

Comment (trendoscope-algorithms) [reply components/ecwid/sources/get-paid-orders/get-paid-orders.mjs:40]: Hi @feyzullah , That is been handled by setting the order status to processing. When we are fetching the new orders, we only fetch those which are in the status awaiting processing.

User also has option to set status to processing if they are handling that in the workflow. Hence, this is given an option which is selected by default.

Comment (feyzullah) [reply components/ecwid/sources/get-paid-orders/get-paid-orders.mjs:40]: Hi @trendoscope-algorithms, As I understand, that logic is implemented on Ecwid side as only items with awaiting status are fetched and those items are the new ones. This one looks good to me but Jorge is on call for PR reviews for this week. Let's request his review.

**Comment by Hassan-Abbas-SQA** [comment]
i have a free account @trendoscope-algorithms does this ECWID needs a paid account to use integrations? like iam unable to login/integrate from that account iam facing this issue 
image

**Comment by trendoscope-algorithms** [comment]
> i have a free account @trendoscope-algorithms does this ECWID needs a paid account to use integrations? like iam unable to login/integrate from that account iam facing this issue image

I am not sure if Ecwid offers free account!! Can you login to your ecwid store and get store id?

**Comment by Hassan-Abbas-SQA** [comment]
> > i have a free account @trendoscope-algorithms does this ECWID needs a paid account to use integrations? like iam unable to login/integrate from that account iam facing this issue image
> 
> I am not sure if Ecwid offers free account!! Can you login to your ecwid store and get store id?

79082003

**Comment by trendoscope-algorithms** [comment]
> 

Not quite sure because the auth module was there before and it just worked for me.
Can you try logging into your Ecwid account first and then try authenticating using pipedream?

**Comment by trendoscope-algorithms** [comment]
> > > i have a free account @trendoscope-algorithms does this ECWID needs a paid account to use integrations? like iam unable to login/integrate from that account iam facing this issue image
> > 
> > 
> > I am not sure if Ecwid offers free account!! Can you login to your ecwid store and get store id?
> 
> 79082003

Not quite sure because the auth module was there before and it just worked for me.
Can you try logging into your Ecwid account first and then try authenticating using pipedream?

**Comment by trendoscope-algorithms** [comment]
> > > > i have a free account @trendoscope-algorithms does this ECWID needs a paid account to use integrations? like iam unable to login/integrate from that account iam facing this issue image
> > > 
> > > 
> > > I am not sure if Ecwid offers free account!! Can you login to your ecwid store and get store id?
> > 
> > 
> > 79082003
> 
> Not quite sure because the auth module was there before and it just worked for me. Can you try logging into your Ecwid account first and then try authenticating using pipedream?

I tried with one of my free account - that is not working for me either. But, I don't see this documented anywhere in ECWID.

I have already setup my paid account anyway which is working ok.

**Comment by Hassan-Abbas-SQA** [comment]
> > > > i have a free account @trendoscope-algorithms does this ECWID needs a paid account to use integrations? like iam unable to login/integrate from that account iam facing this issue image
> > > 
> > > 
> > > I am not sure if Ecwid offers free account!! Can you login to your ecwid store and get store id?
> > 
> > 
> > 79082003
> 
> Not quite sure because the auth module was there before and it just worked for me. Can you try logging into your Ecwid account first and then try authenticating using pipedream?

tried that but no luck 😭
i think i have free account thats why 
image

**Comment by Hassan-Abbas-SQA** [comment]
its working fine 
image

**Review by lcaresia — DISMISSED** [review]

**Comment by trendoscope-algorithms** [comment]
Some other update changed the version of Ecwid component. Just wonder if this is expected?

**Review by lcaresia — APPROVED** [review]
LGTM!

**Pull request merged**